accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,134 @@
+- A list of *packages*, each with:
+  - A *package name*.
+  - A *path* or *location URI*.
+  - A *language version* (major/minor version number)
+
+On top of that, we also add extra metadata which allows us to version the format and attach information about when and how the file was created.
+
+We use JSON for the format. JSON has parsers available for most languages (not all Dart tools are written in Dart) and it is standardized and efficiently parsable.  JSON is *not* easy to write by hand (but doable), and it does not allow comments, but this is a machine generated file that will hardly ever be edited by hand.
+
+## New File Format
+
+The new file will be called `.packages.json` and stored in the same directory as the `.packages` file. The format is JSON text describing is a single JSON object.
+
+That object has, at least, the following properties:
+
+- `apiVersion`: A single integer marking the version of this format, currently `2` (counting `.packages` as the first). Tools should refuse to work on files with a larger version number than the ones they are aware of. Incremental changes will not require increasing the version, so new versions are not expected any time soon, but code should still prepare for it.
+
+- `defaultPackage`: A string containing a package name.
+- `packages`: A list of packages, each with the following properties:
+  - `name`: A string containing a package name.
+  - either `path`: A string containing a relative or absolute path using `/` as separator. Every character other than `/` is used verbatim. If the path starts with a `/` or with a DOS/Windows drive letter (a single letter followed by `:/`), then it is considered absolute. If not, the path is relative to the directory of the `.packages.json` file. 
+  - or `uri`: A URI reference. This is resolved relative to the (likely `file:`) URI referencing the `.package.json` file. If the URI reference has no scheme, or if it has a `file:` scheme, it is recommended to use the `path` instead.
+  - `languageVersion`: Optional. A string containing a language version (major.minor).
+
+Tools should ignore properties that they don't recognize. This will allow us to add more properties in the future without increasing the `apiVersion` number, but not to remove properties or change the meaning of existing properties. 
+
+If we remove a property that tools rely on, change the meaning of an existing property, or add more information that is *required* for resolving and understanding Dart source files, it is a breaking change&mdash;tools expecting the previously available data will no longer understand the whole picture. Such changes will require increasing the `apiVersion` number.
+
+The following properties allows the generator to specify metadata about the file:
+
+- `generated`: A string containing an ISO-8601 UTC timestamp for when the file was generated, using the format: `YYYY-MM-DDTHH:mm:ss.sssZ`.
+- `generator`: An object containing at last the following properties:
```

Comment:
Schema bikeshed: How about combining these under one key:

```json
"generated": {
  "tool": "pub",
  "on": "YYYY-MM-DDTHH:mm:ss.sssZ",
  "version": "1.2.3",
}
```

Reply:
Sure. I was thinking that all generators should have the "generated" time, but not all tools may want the rest. Still, this is probably a good idea.